src/Components/src/Microsoft.AspNetCore.Components.Server/Rendering/HtmlRenderer.cs

```diff
@@ -0,0 +1,207 @@
+            var (segments, _) = RenderFrames(childFrames, 0, childFrames.Count);
+            return (segments, position + frame.ComponentSubtreeLength);
+        }
+
+        private (IEnumerable<string> segments, int nextPosition) RenderElement(ArrayRange<RenderTreeFrame> frames, int position)
+        {
+            var frame = frames.Array[position];
+            IEnumerable<string> elementSegments = new[] { "<", frame.ElementName };
+            var (attributes, afterAttributes) = RenderAttributes(frames, position + 1, frame.ElementSubtreeLength - 1);
+            elementSegments = elementSegments.Concat(attributes);
+            var remainingElements = frame.ElementSubtreeLength + position - afterAttributes;
+            if (remainingElements > 0)
+            {
+                elementSegments = elementSegments.Concat(new[] { ">" });
+                var (children, afterElement) = RenderChildren(frames, afterAttributes, remainingElements);
+                elementSegments = elementSegments.Concat(children).Concat(new[] { "</", frame.ElementName, ">" });
+                Debug.Assert(afterElement == position + frame.ElementSubtreeLength);
+                return (elementSegments, afterElement);
+            }
+            else
+            {
+                elementSegments = elementSegments.Concat(
+                    SelfClosingElements.Contains(frame.ElementName) ?
+                    new[] { " />" } :
+                    new[] { ">", "</", frame.ElementName, ">" });
+                Debug.Assert(afterAttributes == position + frame.ElementSubtreeLength);
+                return (elementSegments, afterAttributes);
+            }
+        }
+
+        private (IEnumerable<string> children, int afterElement) RenderChildren(ArrayRange<RenderTreeFrame> frames, int position, int maxElements)
+        {
+            if (maxElements == 0)
+            {
+                return (Array.Empty<string>(), position);
+            }
+
+            return RenderCore(frames, position, maxElements);
+        }
+
+        private (IEnumerable<string> attributes, int newPosition) RenderAttributes(ArrayRange<RenderTreeFrame> frames, int position, int maxElements)
+        {
+            if (maxElements == 0)
+            {
+                return (Array.Empty<string>(), position);
+            }
+
+            var attributes = new List<string>();
+            for (var i = 0; i < maxElements; i++)
+            {
+                var candidateIndex = position + i;
+                var frame = frames.Array[candidateIndex];
+                if (frame.FrameType != RenderTreeFrameType.Attribute)
+                {
+                    return (attributes, candidateIndex);
+                }
+                if (frame.AttributeValue != null && !ExcludedValueTypes.Any(et => et.IsAssignableFrom(frame.AttributeValue.GetType())))
+                {
+                    attributes.Add($" {frame.AttributeName}={frame.AttributeValue}");
```

Comment:
I think the attribute value should be HTML attribute encoded, and should be enclosed in quotes.

Reply:
Yeah